File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/CancelRuntimeUpgradeInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/** Typed builder for {@code CancelRuntimeUpgrade} instructions. */
public final class CancelRuntimeUpgradeInstruction implements InstructionTemplate {

  public static final String ACTION = "CancelRuntimeUpgrade";
  private static final String ID_HEX = "id_hex";

  private final String idHex;
  private final Map<String, String> arguments;

  private CancelRuntimeUpgradeInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private CancelRuntimeUpgradeInstruction(
      final Builder builder, final Map<String, String> canonicalArguments) {
    this.idHex = builder.idHex;
    this.arguments =
        Collections.unmodifiableMap(
            new LinkedHashMap<>(Objects.requireNonNull(canonicalArguments, "arguments")));
  }

  /** Returns the runtime upgrade id in canonical lowercase hex form. */
  public String idHex() {
    return idHex;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static CancelRuntimeUpgradeInstruction fromArguments(final Map<String, String> arguments) {
    final Builder builder =
        builder()
            .setIdHex(
                GovernanceInstructionUtils.requireHex(
                    require(arguments, ID_HEX), ID_HEX, 32));
    return new CancelRuntimeUpgradeInstruction(builder, new LinkedHashMap<>(arguments));
  }

  private static String require(final Map<String, String> arguments, final String key) {
    final String value = arguments.get(key);
    if (value == null || value.isBlank()) {
      throw new IllegalArgumentException("Instruction argument '" + key + "' is required");
    }
    return value;
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof CancelRuntimeUpgradeInstruction other)) {
      return false;
    }
    return Objects.equals(idHex, other.idHex);
  }

  @Override
  public int hashCode() {
    return Objects.hash(idHex);
  }

  public static final class Builder {
    private String idHex;

    private Builder() {}

    public Builder setIdHex(final String idHex) {
      this.idHex = GovernanceInstructionUtils.requireHex(idHex, "idHex", 32);
      return this;
    }

    public CancelRuntimeUpgradeInstruction build() {
      if (idHex == null || idHex.isBlank()) {
        throw new IllegalStateException("idHex must be provided");
      }
      return new CancelRuntimeUpgradeInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      args.put(ID_HEX, idHex);
      return args;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/CancelRuntimeUpgradeInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

private const val CANCEL_ACTION = "CancelRuntimeUpgrade"
private const val CANCEL_ID_HEX = "id_hex"

/** Typed representation of `CancelRuntimeUpgrade` instructions. */
class CancelRuntimeUpgradeInstruction private constructor(
    val idHex: String,
    private val _arguments: Map<String, String>,
) : InstructionTemplate {

    override val kind: InstructionKind get() = InstructionKind.CUSTOM

    override val arguments: Map<String, String> get() = _arguments

    constructor(idHex: String) : this(
        idHex = GovernanceInstructionUtils.requireHex(idHex, "idHex", 32),
        _arguments = linkedMapOf(
            "action" to CANCEL_ACTION,
            CANCEL_ID_HEX to GovernanceInstructionUtils.requireHex(idHex, "idHex", 32),
        ),
    )

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is CancelRuntimeUpgradeInstruction) return false
        return idHex == other.idHex
    }

    override fun hashCode(): Int = idHex.hashCode()

    companion object {
        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): CancelRuntimeUpgradeInstruction {
            val idHex = GovernanceInstructionUtils.requireHex(
                require(arguments, CANCEL_ID_HEX), CANCEL_ID_HEX, 32,
            )
            return CancelRuntimeUpgradeInstruction(
                idHex = idHex,
                _arguments = LinkedHashMap(arguments),
            )
        }

        private fun require(arguments: Map<String, String>, key: String): String {
            val value = arguments[key]
            require(!value.isNullOrBlank()) { "Instruction argument '$key' is required" }
            return value
        }
    }
}
